File: weather-forecast-web/src/api/UserServices.js
import axios from 'axios';

const USER_SERVICE_URL = process.env.REACT_APP_USER_SERVICE_URL;
// const USER_SERVICE_URL = 'http://192.168.1.40:5000';

const UserServices = {

  signup: async (name, email, password) => {
    try {
      const response = await axios.post(`${USER_SERVICE_URL}/api/signup`, {
        name,
        email,
        password,
      });
      return response.data;
    } catch (error) {
      console.error('Error during signup:', error);
      throw error.response.data;
    }
  },

  signin: async (email, password) => {
    try {
      const response = await axios.post(`${USER_SERVICE_URL}/api/signin`, {
        email,
        password,
      });
      return response.data;
    } catch (error) {
      console.error('Error during signin:', error);
      throw error.response.data;
    }
  },

  tokenIsValid: async (token) => {
    try {
      const response = await axios.post(
        `${USER_SERVICE_URL}/tokenIsValid`,
        {},
        {
          headers: {
            'x-auth-token': token,
          },
        }
      );
      return response.data;
    } catch (error) {
      console.error('Error during token validation:', error);
      throw error.response.data;
    }
  },

  getCurrentUser: async (token) => {
    try {
      const response = await axios.get(`${USER_SERVICE_URL}/`, {
        headers: {
          'x-auth-token': token,
        },
      });
      return response.data;
    } catch (error) {
      console.error('Error fetching current user:', error);
      throw error.response.data;
    }
  },
};

export default UserServices;

File: weather_forecast_app/lib/services/user_services.dart
import 'dart:convert';
import 'package:flutter_dotenv/flutter_dotenv.dart';
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';
import 'package:weather_forecast_app/models/user.dart';
import 'package:weather_forecast_app/provider/user_provider.dart';
import 'package:weather_forecast_app/utils/error_handling.dart';
import 'package:weather_forecast_app/utils/show_snack_bar.dart';

class UserServices {
  final String _baseUrl = dotenv.env['USER_API_URL']!;

  Future<void> signUpUser({
    required BuildContext context,
    required String name,
    required String email,
    required String password,
  }) async {
    try {
      User user = User(
        id: '',
        name: name,
        email: email,
        password: password,
        token: '',
      );

      http.Response res = await http.post(
        Uri.parse('$_baseUrl/api/signup'),
        body: user.toJson(),
        headers: <String, String>{
          'Content-Type': 'application/json; charset=UTF-8'
        },
      );

      httpErrorHandler(
        response: res,
        context: context,
        onSuccess: () {
          showSnackBar(
            context,
            'Account created! Login with the same credentials!',
          );
          Navigator.pop(context);
        },
      );
    } catch (e) {
      showSnackBar(context, e.toString());
    }
  }

  Future<void> signInUser({
    required BuildContext context,
    required String email,
    required String password,
    required bool rememberMe,
  }) async {
    try {
      http.Response res = await http.post(
        Uri.parse('$_baseUrl/api/signin'),
        body: jsonEncode({
          'email': email,
          'password': password,
        }),
        headers: <String, String>{
          'Content-Type': 'application/json; charset=UTF-8'
        },
      );

      httpErrorHandler(
        response: res,
        context: context,
        onSuccess: () async {
          SharedPreferences prefs = await SharedPreferences.getInstance();
          Provider.of<UserProvider>(context, listen: false).setUser(res.body);
          if (rememberMe) {
            await prefs.setString(
                'x-auth-token', jsonDecode(res.body)['token']);
          }
          showSnackBar(
            context,
            'Login successful!',
          );
          Navigator.pop(context);
        },
      );
    } catch (e) {
      showSnackBar(context, e.toString());
    }
  }

  Future<void> signOutUser(BuildContext context) async {
    SharedPreferences prefs = await SharedPreferences.getInstance();
    await prefs.remove('x-auth-token');
    Provider.of<UserProvider>(context, listen: false).clearUser();
  }

  Future<void> getUserData(BuildContext context) async {
    try {
      SharedPreferences prefs = await SharedPreferences.getInstance();
      String? token = prefs.getString('x-auth-token');

      if (token == null) {
        prefs.setString('x-auth-token', '');
      }

      var tokenRes = await http.post(
        Uri.parse('$_baseUrl/tokenIsValid'),
        headers: <String, String>{
          'Content-Type': 'application/json; charset=UTF-8',
          'x-auth-token': token!,
        },
      );

      var response = jsonDecode(tokenRes.body);

      if (response == true) {
        http.Response userRes = await http.get(
          Uri.parse('$_baseUrl/'),
          headers: <String, String>{
            'Content-Type': 'application/json; charset=UTF-8',
            'x-auth-token': token,
          },
        );

        var userProvider = Provider.of<UserProvider>(context, listen: false);
        userProvider.setUser(userRes.body);
      }
    } catch (e) {
      showSnackBar(context, e.toString());
    }
  }
}
